Sources/NIOTransportServices/Datagram/NIOTSDatagramChannel.swift

```diff
@@ -0,0 +1,282 @@
+        }
+
+        switch option {
+        case _ as AutoReadOption:
+            self._options.autoRead = value as! Bool
+            self.readIfNeeded0()
+        case _ as SocketOption:
+            let optionValue = option as! SocketOption
+
+            // SO_REUSEADDR and SO_REUSEPORT are handled here.
+            switch (optionValue.level, optionValue.name) {
+            case (SOL_SOCKET, SO_REUSEADDR):
+                self._reuseAddress = (value as! SocketOptionValue) != Int32(0)
+            case (SOL_SOCKET, SO_REUSEPORT):
+                self._reusePort = (value as! SocketOptionValue) != Int32(0)
+            default:
+                try self.udpOptions.applyChannelOption(option: optionValue, value: value as! SocketOptionValue)
+            }
+        case _ as WriteBufferWaterMarkOption:
+            if self._backpressureManager.writabilityChanges(whenUpdatingWaterMarks: value as! WriteBufferWaterMark) {
+                self.pipeline.fireChannelWritabilityChanged()
+            }
+        case is NIOTSEnablePeerToPeerOption:
+            self._enablePeerToPeer = value as! NIOTSEnablePeerToPeerOption.Value
+        default:
+            fatalError("option \(type(of: option)).\(option) not supported")
+        }
+    }
+
+    public func getOption<Option: ChannelOption>(_ option: Option) -> EventLoopFuture<Option.Value> {
+        if eventLoop.inEventLoop {
+            let promise: EventLoopPromise<Option.Value> = eventLoop.makePromise()
+            executeAndComplete(promise) { try getOption0(option: option) }
+            return promise.futureResult
+        } else {
+            return eventLoop.submit { try self.getOption0(option: option) }
+        }
+    }
+
+    func getOption0<Option: ChannelOption>(option: Option) throws -> Option.Value {
+        self.eventLoop.assertInEventLoop()
+
+        guard !self.closed else {
+            throw ChannelError.ioOnClosedChannel
+        }
+
+        switch option {
+        case _ as AutoReadOption:
+            return self._options.autoRead as! Option.Value
+        case _ as SocketOption:
+            let optionValue = option as! SocketOption
+
+            // SO_REUSEADDR and SO_REUSEPORT are handled here.
+            switch (optionValue.level, optionValue.name) {
+            case (SOL_SOCKET, SO_REUSEADDR):
+                return Int32(self._reuseAddress ? 1 : 0) as! Option.Value
+            case (SOL_SOCKET, SO_REUSEPORT):
+                return Int32(self._reusePort ? 1 : 0) as! Option.Value
+            default:
+                return try self.udpOptions.valueFor(socketOption: optionValue) as! Option.Value
+            }
+        case _ as WriteBufferWaterMarkOption:
+            return self._backpressureManager.waterMarks as! Option.Value
+        case is NIOTSEnablePeerToPeerOption:
+            return self._enablePeerToPeer as! Option.Value
+        default:
+            fatalError("option \(type(of: option)).\(option) not supported")
+        }
+    }
```

Comment:
I don't think we should need this implemented here: it should be able to be implemented on `StateManagedNWConnectionChannel`. It'll require defining a typealias for the `udpOptions` (and `tcpOptions` for `NWConnectionChannel`), but that's fine.